private/bufpkg/bufcheck/buflint/internal/buflintvalidate/cel.go

```diff
@@ -118,19 +99,120 @@ func validateCELCompilesField(
 	if err != nil {
 		return err
 	}
-	for i, celConstraint := range fieldConstraints.GetCel() {
-		celLocation := field.OptionExtensionLocation(
-			validate.E_Field,
-			celFieldTagInFieldConstraints,
-			int32(i),
-		)
-		checkCel(celEnv, celConstraint, func(format string, args ...interface{}) {
-			add(field, celLocation, nil, format, args...)
-		})
-	}
+	checkCEL(
+		celEnv,
+		fieldConstraints.GetCel(),
+		fmt.Sprintf("field %q", adder.fieldName()),
+		fmt.Sprintf("Field %q", adder.fieldName()),
+		adder.getFieldRuleName(celFieldNumberInFieldConstraints),
+		func(index int, format string, args ...interface{}) {
+			adder.addForPathf(
+				[]int32{celFieldNumberInFieldConstraints, int32(index)},
+				format,
+				args...,
+			)
+		},
+	)
 	return nil
 }
 
+func checkCEL(
+	celEnv *cel.Env,
+	celConstraints []*validate.Constraint,
+	parentName string,
+	parentNameCapitalized string,
+	celName string,
+	add func(int, string, ...interface{}),
+) {
+	idToConstraintIndices := make(map[string][]int, len(celConstraints))
+	for i, celConstraint := range celConstraints {
+		if celID := celConstraint.GetId(); celID != "" {
+			for _, char := range celID {
+				if 'a' <= char && char <= 'z' {
+					continue
+				} else if 'A' <= char && char <= 'Z' {
+					continue
+				} else if '0' <= char && char <= '9' {
+					continue
+				} else if char == '_' || char == '-' || char == '.' {
+					continue
+				}
+				add(
+					i,
+					"%s has an invalid %s.id, an id must contain only alphanumerics and '.', '_' and '-'.",
+					parentNameCapitalized,
+					celName,
+				)
+				break
+			}
+			idToConstraintIndices[celID] = append(idToConstraintIndices[celID], i)
+		} else {
+			add(i, "%s has an empty %s.id.", parentNameCapitalized, celName)
+		}
+		if len(strings.TrimSpace(celConstraint.Expression)) == 0 {
+			add(i, "%s has an empty %s.expression.", parentNameCapitalized, celName)
+			continue
+		}
+		ast, compileIssues := celEnv.Compile(celConstraint.Expression)
+		switch {
+		case ast.OutputType().IsAssignableType(cel.BoolType):
+			if celConstraint.Message == "" {
+				add(
+					i,
+					"%s has an empty %s.message for an expression that evaluates to a boolean.",
+					parentNameCapitalized,
+					celName,
+				)
+			}
+		case ast.OutputType().IsAssignableType(cel.StringType):
+			if celConstraint.Message != "" {
+				add(
+					i,
+					"%s has a %s with a string expression and a redundant message.",
```

Comment:
`%s has a %s with an expression that evaluates to a string, and also has a message. The message is redundant - since the expression evaluates to a string, its result will be printed instead of the message, so the message should be removed.`

Reply:
resolved in [6a37deb](https://github.com/bufbuild/buf/pull/2505/commits/6a37debb7800c2b4cd1dee0520f712038648a63a)